Fix py_nomainwindow_SegmentationsModuleTest2 on Linux

This test is failing on Linux with either VTK8 or VTK9

(Incidentally, it is also failing on macOS because of path to long in shared library. See [here](http://slicer.cdash.org/test/16388347)) 

## Steps to reproduce

```
ctest -R py_nomainwindow_SegmentationsModuleTest2 -V 
[...]
test 407
    Start 407: py_nomainwindow_SegmentationsModuleTest2

407: Test command: /path/to/Slicer-VTK8-rwdi/Slicer-build/Slicer "--no-splash" "--testing" "--disable-cli-modules" "--no-main-window" "--additional-module-paths" "/path/to/Slicer-VTK8-rwdi/Slicer-build/lib/Slicer-4.13/qt-loadable-modules" "/path/to/Slicer-VTK8-rwdi/Slicer-build/lib/Slicer-4.13/qt-scripted-modules" "--additional-module-paths" "/path/to/Slicer-VTK8-rwdi/Slicer-build/lib/Slicer-4.13/qt-scripted-modules" "/path/to/Slicer-VTK8-rwdi/Slicer-build/lib/Slicer-4.13/cli-modules" "/path/to/Slicer-VTK8-rwdi/Slicer-build/lib/Slicer-4.13/qt-loadable-modules" "--python-code" "import slicer.testing; slicer.testing.runUnitTest(['/path/to/Slicer-VTK8-rwdi/Slicer-build/Modules/Loadable/Segmentations/Testing/Python', '/path/to/Slicer/Modules/Loadable/Segmentations/Testing/Python'], 'SegmentationsModuleTest2')"
407: Test timeout computed to be: 1500
407: When loading module  "CropVolume" , the dependency "ResampleScalarVectorDWIVolume" failed to be loaded.
407: qSlicerSequencesModulePrivate::addToolBar: no main window is available, toolbar is not added
407: -------------------------------------------
407: path: ['/path/to/Slicer-VTK8-rwdi/Slicer-build/Modules/Loadable/Segmentations/Testing/Python', '/path/to/Slicer/Modules/Loadable/Segmentations/Testing/Python']
407: testname: SegmentationsModuleTest2
407: -------------------------------------------
407: runTest (slicer.ScriptedLoadableModule.ScriptedLoadableModuleTest) ... No test is defined in ScriptedLoadableModuleTest
407: ok
407: Requested file has been found and checksum is OK: /tmp/Slicer-tmp-jcfr/SegmentationsModuleTest/TinyPatient_Seg.zip
407: Unzipping /tmp/Slicer-tmp-jcfr/SegmentationsModuleTest/TinyPatient_Seg.zip into /tmp/Slicer-tmp-jcfr/SegmentationsModuleTest successful
407: Loaded volume from file: /tmp/Slicer-tmp-jcfr/SegmentationsModuleTest/TinyPatient_Seg/TinyPatient_CT.nrrd. Dimensions: 10x10x10. Number of components: 1. Pixel type: int.
407: 
407: 
407: "Volume" Reader has successfully read the file "/tmp/Slicer-tmp-jcfr/SegmentationsModuleTest/TinyPatient_Seg/TinyPatient_CT.nrrd" "[0.02s]"
407: Multiple layer editing successful
407: 5 islands created (2 ignored)
407: 3 islands created (4 ignored)
407: 5 islands created (2 ignored)
407: Running test on margin effect
407: Testing shared labelmaps
407: Running margin effect with data type: 2, and overwriteMode 0
407: Running margin effect with data type: 2, and overwriteMode 2
407: Running margin effect with data type: 15, and overwriteMode 0
407: Running margin effect with data type: 15, and overwriteMode 2
407: Running margin effect with data type: 3, and overwriteMode 0
407: Running margin effect with data type: 3, and overwriteMode 2
407: Running margin effect with data type: 4, and overwriteMode 0
407: Running margin effect with data type: 4, and overwriteMode 2
407: Running margin effect with data type: 5, and overwriteMode 0
407: Running margin effect with data type: 5, and overwriteMode 2
407: Running margin effect with data type: 6, and overwriteMode 0
407: Running margin effect with data type: 6, and overwriteMode 2
407: Running margin effect with data type: 7, and overwriteMode 0
407: Running margin effect with data type: 7, and overwriteMode 2
407: Running margin effect with data type: 8, and overwriteMode 0
407: test_SegmentationsModuleTest2 (SegmentationsModuleTest2.SegmentationsModuleTest2) ... FAIL
407: 
407: ======================================================================
407: FAIL: test_SegmentationsModuleTest2 (SegmentationsModuleTest2.SegmentationsModuleTest2)
407: ----------------------------------------------------------------------
407: Traceback (most recent call last):
407:   File "/path/to/Slicer/Modules/Loadable/Segmentations/Testing/Python/SegmentationsModuleTest2.py", line 44, in test_SegmentationsModuleTest2
407:     self.TestSection_MarginEffects()
407:   File "/path/to/Slicer/Modules/Loadable/Segmentations/Testing/Python/SegmentationsModuleTest2.py", line 295, in TestSection_MarginEffects
407:     self.runMarginEffect(segment1, segment2, dataType, self.segmentEditorNode.OverwriteAllSegments)
407:   File "/path/to/Slicer/Modules/Loadable/Segmentations/Testing/Python/SegmentationsModuleTest2.py", line 343, in runMarginEffect
407:     self.checkSegmentVoxelCount(1, 1)
407:   File "/path/to/Slicer/Modules/Loadable/Segmentations/Testing/Python/SegmentationsModuleTest2.py", line 248, in checkSegmentVoxelCount
407:     self.assertEqual(imageStat.GetVoxelCount(), expectedVoxelCount)
407: AssertionError: 0 != 1
407: 
407: ----------------------------------------------------------------------
407: Ran 2 tests in 0.425s
407: 
407: FAILED (failures=1)
407: Traceback (most recent call last):
407:   File "<string>", line 1, in <module>
407:   File "/path/to/Slicer-VTK8-rwdi/Slicer-build/bin/Python/slicer/testing.py", line 26, in runUnitTest
407:     exitFailure()
407:   File "/path/to/Slicer-VTK8-rwdi/Slicer-build/bin/Python/slicer/testing.py", line 11, in exitFailure
407:     raise Exception(message)
407: Exception
407: vtkDebugLeaks has found no leaks.
1/1 Test #407: py_nomainwindow_SegmentationsModuleTest2 ...***Failed    3.67 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   3.71 sec

The following tests FAILED:
	407 - py_nomainwindow_SegmentationsModuleTest2 (Failed)
```

## Expected behavior

Passing test

## Environment
- Slicer version: Slicer-4.??.?-YYYY-MM-DD
- Operating system: Windows / Linux / Mac + which version

--- a/Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx
+++ b/Libs/vtkSegmentationCore/vtkOrientedImageDataResample.cxx
@@ -1877,36 +1877,24 @@ void vtkOrientedImageDataResample::CastImageForValue(vtkOrientedImageData* image
 
   if (typeIsSigned)
     {
-    if (value > VTK_FLOAT_MAX || value < VTK_FLOAT_MIN)
+    if (value > VTK_INT_MAX || value < VTK_INT_MIN)
       {
       scalarType = VTK_DOUBLE;
       }
-    else if (value > VTK_LONG_MAX || value < VTK_LONG_MIN)
-      {
-      scalarType = VTK_FLOAT;
-      }
-    else if (value > VTK_INT_MAX || value < VTK_INT_MIN)
-      {
-      scalarType = VTK_LONG;
-      }
     else if (value > VTK_SHORT_MAX || value < VTK_SHORT_MIN)
+      {
+      scalarType = VTK_INT;
+      }
+    else if (value > VTK_SIGNED_CHAR_MAX || value < VTK_SIGNED_CHAR_MIN)
       {
       scalarType = VTK_SHORT;
       }
     }
   else
     {
-    if (value > VTK_FLOAT_MAX)
+    if (value > VTK_UNSIGNED_INT_MAX)
       {
       scalarType = VTK_DOUBLE;
-      }
-    else if (value > static_cast<double>(VTK_UNSIGNED_LONG_MAX))
-      {
-      scalarType = VTK_FLOAT;
-      }
-    else if (value > VTK_UNSIGNED_INT_MAX)
-      {
-      scalarType = VTK_UNSIGNED_LONG;
       }
     else if (value > VTK_UNSIGNED_SHORT_MAX)
       {

--- a/Modules/Loadable/Segmentations/Testing/Python/SegmentationsModuleTest2.py
+++ b/Modules/Loadable/Segmentations/Testing/Python/SegmentationsModuleTest2.py
@@ -280,9 +280,9 @@ class SegmentationsModuleTest2(unittest.TestCase):
       vtk.VTK_UNSIGNED_SHORT,
       vtk.VTK_INT,
       vtk.VTK_UNSIGNED_INT,
-      vtk.VTK_LONG,
-      vtk.VTK_UNSIGNED_LONG,
-      vtk.VTK_FLOAT,
+      #vtk.VTK_LONG, # On linux, VTK_LONG has the same size as VTK_LONG_LONG. This causes issues in vtkImageThreshold.
+      #vtk.VTK_UNSIGNED_LONG, See https://github.com/Slicer/Slicer/issues/5427
+      #vtk.VTK_FLOAT, # Since float can't represent all int, we jump straight to double.
       vtk.VTK_DOUBLE,
       #vtk.VTK_LONG_LONG, # These types are unsupported in ITK
       #vtk.VTK_UNSIGNED_LONG_LONG,

